pkg/reconciler/reconciler.go

```diff
@@ -537,10 +556,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
 	}
 	u.UpdateStatus(updater.EnsureCondition(conditions.Irreconcilable(corev1.ConditionFalse, "", "")))
 
-	for _, h := range r.preHooks {
-		if err := h.Exec(obj, vals, log); err != nil {
-			log.Error(err, "pre-release hook failed")
-		}
+	r.extensions.loggerInto(log)
```

Comment:
I don't think I understand why injecting loggers is needed, especially at runtime on each reconciliation. If it's about compatibility with the old hook interface, any retrofitting of interfaces should IMHO happen exactly once, namely at registration time. Meaning that when you call `WithPreHook`, the hook function is converted into an extension and that extension is then registered. Subsequent logic will treat this extension as every other extension.

Reply:
Well, Joe suggested to use an injection API as it is currently being done within controller-runtime. And the thing with the `Reconcile()` method is that it currently derives a new logger from the existing one with a special `WithValues()` configuration based on the `req` parameter of the `Reconcile()` function. :-/